pkg/sources/errors.go

```diff
@@ -0,0 +1,32 @@
+package sources
+
+import (
+	"sync"
+	"sync/atomic"
+)
+
+// ScanErrors is used to collect errors encountered while scanning.
+// It ensures that errors are collected in a thread-safe manner.
+type ScanErrors struct {
+	count  uint64
+	mu     sync.Mutex
+	errors []error
+}
```

Comment:
**Nit:** `count` is `len(errors)`, could we just use that? I feel slightly better having a single source of truth for the number of errors. Also I think `mu` could be a `RWMutex`

Reply:
So the idea here was to encapsulate all details around the errors within this object. Additionally errors is not accessible since it's a private field on this struct. All the sources that use this object should not have direct access to any of the internals. So while the single source of truth is the len(errors) it is not accessible except through the `Count()` method.

Oh, good point, will use RWMutex instead.